### PR TITLE
crystal: 1.9 -> 1.11

### PR DESCRIPTION
**Pull request by sefidel**
## Description of changes

Bump versions.

Changes: https://github.com/crystal-lang/crystal/compare/1.9.2...1.11.2
Changelog: https://github.com/crystal-lang/crystal/blob/master/CHANGELOG.md#1112-2024-01-18

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by sefidel** [comment]
> please include link to relase notes, changelog and/or diff in the commit message ( and PR )

Oops, can't believe I missed that! Fixed.

**Comment by kirillrdy** [comment]
Result of `nixpkgs-review pr 286273` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>ameba</li>
    <li>collision</li>
    <li>crystalline</li>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal</li>
    <li>crystal.bin</li>
    <li>crystal.lib</li>
    <li>crystal2nix</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>

**Comment by sefidel** [comment]
Most of them are Gio errors which is addressed on PRs below:

ameba: https://github.com/NixOS/nixpkgs/pull/286275
collision: https://github.com/NixOS/nixpkgs/pull/282286
crystalline: https://github.com/NixOS/nixpkgs/pull/286276
rtfm: broken since 2023-12-29, [Hydra logs](https://hydra.nixos.org/build/248174386/nixlog/2)

**Review by hughobrien — APPROVED** [review]

**Comment by hughobrien** [comment]
Result of `nixpkgs-review pr 286273` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>ameba</li>
    <li>collision</li>
    <li>crystalline</li>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>18 packages built:</summary>
  <ul>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal</li>
    <li>crystal.bin</li>
    <li>crystal.lib</li>
    <li>crystal2nix</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>

**Comment by kirillrdy** [comment]
@sefidel since there are breaking changes, can you please add an entry to nixos release notes

**Comment by peterhoeg** [comment]
@sefidel, can I trouble you to bring in the other PRs you have created addressing crystal programs into this? As separate commits of course, but this way we can be sure that everything works and gets merged in one go.

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 1]
```diff
@@ -228,6 +228,9 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
   - The `erlang_node_short_name`, `erlang_node_name`, `port` and `options` configuration parameters are gone, and have been replaced with an `environment` parameter.
     Use the appropriate [environment variables](https://hexdocs.pm/livebook/readme.html#environment-variables) inside `environment` to configure the service instead.
 
+- The `crystal` package has been updated to 1.11.x, which has some breaking changes.
```
Comment (kirillrdy) [inline nixos/doc/manual/release-notes/rl-2405.section.md:231]: :+1:

**Comment by sefidel** [comment]
> bring in the other PRs you have created addressing crystal programs into this

@peterhoeg I've merged https://github.com/NixOS/nixpkgs/pull/286275 and https://github.com/NixOS/nixpkgs/pull/286276 to this PR.

**Comment by peterhoeg** [comment]
@GrahamcOfBorg build ameba crystalline

**Comment by sefidel** [comment]
Squashed duplicate commit.

**Comment by sefidel** [comment]
Result of `nixpkgs-review pr 286273` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>collision</li>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>ameba</li>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal</li>
    <li>crystal.bin</li>
    <li>crystal.lib</li>
    <li>crystal2nix</li>
    <li>crystalline</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>

File: nixos/doc/manual/release-notes/rl-2405.section.md (hunk 1 above)
Comment (soupglasses) [reply nixos/doc/manual/release-notes/rl-2405.section.md:231]: Maybe link to https://github.com/crystal-lang/crystal/blob/master/CHANGELOG.md#1110-2024-01-08 and say to look at it if you are affected?

**Comment by sund3RRR** [comment]
I think we need to figure something out with [crystal2nix](https://github.com/nix-community/crystal2nix). Crystal dependencies are a pain in the ass when packing through `shards.nix`. This is due to the fact that the state of the crystal project dependencies is not final at the time of unpacking. However, `crystal2nix` puts them in the  `/nix/store` and forbids them to change their state during the build of the project.
1) `shards.yml` file may contain post installation scripts;
2) crystal sources can have macros and write some data at the compilation time to the dependencies folder;
3) unpatched [gi-crystal](https://github.com/hugopl/gi-crystal), which had to be packaged in nix, generally generates bindings in its own folder and instantly breaks the build, because `/nix/store` is read-only.

Right now I have a [collision PR](https://github.com/NixOS/nixpkgs/pull/282286) hanging, but I have no idea how to solve encountered problems correctly, it seems impossible, since `blake3-cr` should compile the library, but it cannot do this while in the `/nix/store` for obvious reasons.

The `rtfm` package is broken due to the `gi-crystal` update. The `gi-crystal` version is explicitly registered in `shards.lock`, however, it is not possible to use an unpatched `gi-crystal` in `shards.nix`, since it will not be able to write bindings to `/nix/store`.

**Comment by will** [comment]
Result of `nixpkgs-review pr 286273` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>collision</li>
  </ul>
</details>
<details>
  <summary>19 packages built:</summary>
  <ul>
    <li>ameba</li>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal</li>
    <li>crystal.bin</li>
    <li>crystal.lib</li>
    <li>crystal2nix</li>
    <li>crystalline</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
  </ul>
</details>

**Comment by will** [comment]
From what I can tell the only thing that failed to build will be addressed by https://github.com/NixOS/nixpkgs/pull/282286 which has the last messsage:

> I will make changes to the current PR as soon as crystal's PR is merged.

So this PR is probably good to go now?

**Comment by sund3RRR** [comment]
> From what I can tell the only thing that failed to build will be addressed by #282286 which has the last messsage:
> 
> > I will make changes to the current PR as soon as crystal's PR is merged.
> 
> So this PR is probably good to go now?

I talked about this [PR](https://github.com/NixOS/nixpkgs/pull/289266) if it matters

**Comment by sund3RRR** [comment]
I want to fix all the projects related to this, but no one review my PR so I'm sad

**Comment by hughobrien** [comment]
Result of `nixpkgs-review pr 286273` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>collision</li>
    <li>rtfm</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>ameba</li>
    <li>amqpcat</li>
    <li>blahaj</li>
    <li>crystal</li>
    <li>crystal.bin</li>
    <li>crystal.lib</li>
    <li>crystal2nix</li>
    <li>crystalline</li>
    <li>gi-crystal</li>
    <li>icr</li>
    <li>invidious</li>
    <li>kakoune-cr</li>
    <li>lucky-cli</li>
    <li>mint</li>
    <li>oq</li>
    <li>scry</li>
    <li>shards (shards_0_17)</li>
    <li>thicket</li>
    <li>tijolo</li>
    <li>tmuxPlugins.fingers</li>
  </ul>
</details>

**Pull request merged**